Integrate DTS emulator with workspace view

philliphoff opened the pull request:
Adds a DTS emulator node to the Workspace view of the Azure Resources explorer, which will show running DTS emulator instances.  This node offers the ability to start new emulator instances.  Emulator nodes, like in the Azure Resources view, offers the abilities:

 - Copy connections strings
 - Open the corresponding task hub dashboard
 - Shutdown the emulator

The DTS settings have been updated to allow customization (i.e. registry, image, tag) of the emulator container image.  Emulators started by the VS Code instance will be shutdown when the instance is closed.  (Emulators started outside of the VS Code instance will *not* be shutdown.)

Note: the emulator features are not hidden behind the "preview features" toggle as they are (as of now) available to all.

philliphoff commented:
@nturinski Thanks for taking a look.

> Phil, would it be possible to add a command to install/uninstall the emulator similar to Azure Functions Core Tools?

The emulator is just a Docker container; there's nothing to install or uninstall really (save pulling/removing an image).  I'll think about whether we can show/hide the workspace node based on the existence of the image (and then have a command that just pulls it).  I feel that's not terribly discoverable, though.

> When you activate the extension, it immediately throws an error as well:

Well that's not good; I'll try to repro and resolve that.

nturinski commented:
> The emulator is just a Docker container; there's nothing to install or uninstall really (save pulling/removing an image). I'll think about whether we can show/hide the workspace node based on the existence of the image (and then have a command that just pulls it). I feel that's not terribly discoverable, though.

I think if the node could instruct users on where/how to get the image, that could be helpful. Like a disabled/enabled state? That way we could still have the discoverability regardless.

philliphoff commented:
> I think if the node could instruct users on where/how to get the image, that could be helpful. Like a disabled/enabled state? That way we could still have the discoverability regardless.

@nturinski The image will be auto-pulled if/when the user attempts to start the container, so there's nothing they need to do (aside from Docker being running).

That said, I've added some robustness to the Docker management logic, and on errors the DTS Emulators node will show a warning node with the error (e.g. "is Docker installed and running?"). I've also updated the root node to auto-expand *only* if emulators are already running.  This *should* make the feature a little less "in your face" if you're not actually using it.

alexweininger commented:
Would you want to try to start the Docker runtime for the user if it's not already running? I know there are other runtimes, but I think most devs still use Docker. Mostly suggesting it because I was using the dev containers feature in VS Code today and it auto started Docker desktop for me so it's fresh in my mind.

alexweininger commented:
Non-blocking feedback here:

<img width="459" alt="image" src="https://github.com/user-attachments/assets/b5cb8969-bd89-4053-b402-a6e23b155234" />
<img width="446" alt="image" src="https://github.com/user-attachments/assets/1246a59a-5013-48a4-bc6e-f91b00a253c4" />

Since there aren't that many actions available on the emulator tree item and the task hub item, I think it makes sense to expose the "big" ones as buttons on the tree item itself. Similar to the buttons we have on the Functions local project tree item:
<img width="470" alt="Screenshot 2025-03-19 at 5 13 23 PM" src="https://github.com/user-attachments/assets/cb60eaba-d011-43c1-a069-cd295b44e0e1" />

I think at least a stop/start button on the emulator item would make a lot of sense. And an open in dashboard button on the task hub. I figure we might as well make these things as discoverable as possible.

philliphoff commented:
> Would you want to try to start the Docker runtime for the user if it's not already running? I know there are other runtimes, but I think most devs still use Docker. Mostly suggesting it because I was using the dev containers feature in VS Code today and it auto started Docker desktop for me so it's fresh in my mind.

@alexweininger I talked to the Docker extension team and they, themselves, do not automatically start the container runtime so I'm hesitant to try to do that in this case.  (Just trying to understand *which* runtime is in use and how to start it would be a significant concern--I'm using the Docker extension client library to abstract away the runtime being used, so the extension doesn't actually know.)

philliphoff commented:
> I think at least a stop/start button on the emulator item would make a lot of sense. And an open in dashboard button on the task hub. I figure we might as well make these things as discoverable as possible.

@alexweininger Those are good ideas and I've added inline commands for opening the taskhub dashboard, as well as for starting and stopping emulator instances.

philliphoff commented:
@alexweininger Bump.  (Not in any hurry, just don't want to forget about it myself.)

nturinski commented on src/tree/durableTaskScheduler/DurableTaskSchedulerEmulatorsWorkspaceResourceModel.ts:
```diff
@@ -0,0 +1,70 @@
+
+    private async emulatorsExist(): Promise<boolean> {
+        // Used any cached result that might exist...
+        this.getEmulatorsTask ??= this.emulatorClient.getEmulators();
```
Is there any action that the user could take to change what `getEmulators` would output during a session? 

I ask because `getEmulatorsTask` never gets set to undefined, so this list would never get updated even if the user stopped an emulator. But I may be misunderstanding what "exists" means in this case.

philliphoff replied:
`getEmulatorsTask` is updated when the node is asked for children.  That said, the logic takes advantage of the fact that VS Code will refresh/recreate the parent node entirely in the event of user action (e.g. manually refreshing the window, or starting or stopping an emulator).  (See the provider's subscription to emulator changes.). The observation was that VS Code tended to call getTreeItem() quite often, for the same node, so the thought was to try to eliminate redundant calls to relatively slow container commands.

nturinski commented on src/tree/durableTaskScheduler/DurableTaskSchedulerErrorWorkspaceResourceModel.ts:
```diff
@@ -0,0 +1,26 @@
+        private readonly error: string) {
+    }
+
+    getChildren(): ProviderResult<DurableTaskSchedulerWorkspaceResourceModel[]> {
```
Should this item have any children? I believe it'll make it expandable, but then nothing would show below which is an odd UI experience.

philliphoff replied:
The `TreeItemCollapsableState` controls whether the expansion UX is shown (and, by default, not).  That said, I'll update the model interface to make this optional so it can be omitted, and be clearer, for those that don't have children.

nturinski replied:
I think you missed this one when removing `getChildren` definitions.

philliphoff replied:
Bah!  Sorry about that.

nturinski reviewed (commented):
There were a couple of times get children returned an empty array or the context values were joined with a space. I didn't feel like I needed to comment on every single one, but if you'd like me to just so it's easier to identify where those all are, I can do that.

philliphoff commented on package.json:
```diff
@@ -797,8 +823,33 @@
+                },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.startEmulator",
+                    "when": "view =~ /(azureWorkspace|azureFocusView)/ && viewItem =~ /azFunc.dts.emulators/",
```
I think I was just following the pattern found elsewhere.  I'll make it `azureWorkspace` specific.

philliphoff commented on package.json:
```diff
@@ -880,6 +935,14 @@
                     "when": "never"
                 },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.startEmulator",
```
Good point. I'll keep it visible.

philliphoff commented on package.json:
```diff
@@ -997,6 +1060,21 @@
+                    "azureFunctions.durableTaskScheduler.emulatorTag": {
+                        "type": "string",
+                        "description": "%azureFunctions.durableTaskScheduler.emulatorTag%",
+                        "default": "v0.0.4"
```
As of last week, we do now have a rolling `latest` tag. I'll update the default.

philliphoff commented on src/tree/durableTaskScheduler/DurableTaskSchedulerEmulatorClient.ts:
```diff
@@ -0,0 +1,122 @@
+}
+
+function getEmulatorFullTag(): ImageFullTag {
+    const configuration = workspace.getConfiguration('azureFunctions');
```
Nice! Will use them.

philliphoff commented on src/tree/durableTaskScheduler/DurableTaskSchedulerEmulatorsWorkspaceResourceModel.ts:
```diff
@@ -0,0 +1,70 @@
+        try {
+            await this.getEmulatorsTask;
+
+            const emulators = await this.getEmulatorsTask;
```
Ah, yes, that is redundant.

nturinski commented on package.json:
```diff
@@ -765,14 +786,19 @@
                 },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.copyEmulatorConnectionString",
+                    "when": "view =~ /(azureWorkspace|azureFocusView)/ && viewItem =~ /azFunc.dts.emulatorInstance/",
```
This probably doesn't need `azureFocusView` in the view regex.

philliphoff replied:
Updated.

nturinski commented on package.json:
```diff
@@ -797,8 +823,33 @@
+                },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.startEmulator",
+                    "when": "view =~ /(azureWorkspace|azureFocusView)/ && viewItem =~ /azFunc.dts.emulators/",
```
This probably doesn't need `azureFocusView` in the view regex.

philliphoff replied:
Updated.

nturinski commented on package.json:
```diff
@@ -797,8 +823,33 @@
+                },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.stopEmulator",
+                    "when": "view =~ /(azureWorkspace|azureFocusView)/ && viewItem =~ /azFunc.dts.emulatorInstance/",
```
This probably doesn't need `azureFocusView` in the view regex.

philliphoff replied:
Updated.

nturinski commented on package.json:
```diff
@@ -797,8 +823,33 @@
+                },
+                {
+                    "command": "azureFunctions.durableTaskScheduler.stopEmulator",
+                    "when": "view =~ /(azureWorkspace|azureFocusView)/ && viewItem =~ /azFunc.dts.emulatorInstance/",
```
This probably doesn't need `azureFocusView` in the view regex.

philliphoff replied:
Updated.

nturinski reviewed (commented):
Overall looks good to me. Let me know whenever you're done-- any approval I do now is just going to get stale.

nturinski commented:
@philliphoff I'm going to hold off on merging this for just a couple of days. We had CTI run a release test pass this week and we were planning to release today or Monday. 

We'll release again soon including this and other DTS changes.

philliphoff commented:
No worries!  Thanks for taking the time to review!
